applications/FluidDynamicsHydraulicsApplication/custom_utilities/hydraulic_fluid_auxiliary_utilities.cpp

```diff
@@ -0,0 +1,322 @@
+                KRATOS_ERROR_IF(n_nodes != 3) << "This function only supports Triangle3D3N geometries." << std::endl;
+
+                // Loop through pairs of nodes to identify unique edges
+                for (IndexType i = 0; i < n_nodes - 1; ++i)
+                {
+                    const IndexType i_id = r_geom[i].Id();
+                    for (IndexType j = i + 1; j < n_nodes; ++j)
+                    {
+                        const IndexType j_id = r_geom[j].Id();
+
+                        // Ensure consistent ordering of node pairs for uniqueness
+                        if (i_id < j_id) {
+                            aux_pair = std::make_pair<int, int>(i_id, j_id);
+                        } else{
+                            aux_pair = std::make_pair<int, int>(j_id, i_id);
+                        }
+
+                        // Check if the edge is already in the map
+                        auto found = edges_map.find(aux_pair);
+                        // If not, create a new entry in the map
+                        if (found == edges_map.end())
+                        {
+                            EdgeDataContainer edge_data;
+                            edge_data.pNodeI = i_id < j_id ? r_geom(i) : r_geom(j);
+                            edge_data.pNodeJ = i_id < j_id ? r_geom(j) : r_geom(i);
+                            edge_data.NumberOfRepetitions = 1;
+                            edges_map.insert(std::make_pair(aux_pair, edge_data));
+                        }
+                        // If the edge is already in the map, update the repetition count
+                        else
+                        {
+                            auto &r_edge_data = found->second;
+                            r_edge_data.NumberOfRepetitions += 1;
+                        }
+                    }
+                }
+            }
+        }
+
+        for (auto it = edges_map.begin(); it != edges_map.end();)
+        {   // Delete all edges that have more than one repetition, since they are not part of the perimeter
+            const auto& r_edge_data = it->second;
+            if (r_edge_data.NumberOfRepetitions > 1) {
+                it = edges_map.erase(it);
+            } else {
+                ++it;
+            }
+        }
+        // Calculate de distance of each edge belonging to the perimeter.
+        for (auto it = edges_map.begin(); it != edges_map.end(); ++it) {
+            const auto& r_edge_data = it->second;
+            const double distance_value_j = distance_getter(*(r_edge_data.pNodeJ), rDistanceVariable);
+            const double distance_value_i = distance_getter(*(r_edge_data.pNodeI), rDistanceVariable);
+            // Case 1: The edge is completly wetted
+            if ((distance_value_i<0.0) && (distance_value_j<0.0) ){
+                const array_1d<double, 3> edge_vector = r_edge_data.pNodeJ->Coordinates() - r_edge_data.pNodeI->Coordinates();
```

Comment:
Here you're allocating one vector for each edge. You can optimize it by allocating one once outside and using it in here.